### ISSUE TITLE
BUG: Cannot parse a specific Android Vector Drawable

### ISSUE THREAD

**Issue by alesimula** [zathras/jovial_svg#14]
Hello, I was trying to parse this simple vector drawable:

```
<?xml version="1.0" encoding="utf-8"?>
<vector xmlns:android="http://schemas.android.com/apk/res/android" android:height="108dip" android:width="108dip" android:viewportWidth="432" android:viewportHeight="432">
    <path android:fillColor="#ff1877f2" android:pathData="M231.7,129.6c-14,3.4 -26.2,12.8 -31.7,24.6 -5.3,11.4 -6,16.1 -6,40l0,21.8 -18.5,-0 -18.5,-0 0,21 0,21 18.5,-0 18.5,-0 0,87 0,87 22,-0 22,-0 0,-87 0,-87 17.3,-0 17.3,-0 2.8,-19.2c1.6,-10.6 3.1,-20.1 3.4,-21 0.4,-1.7 -1,-1.8 -20.3,-1.8l-20.7,-0 0.4,-16.3c0.3,-12.4 0.9,-17.1 2.2,-20.2 4.2,-9.7 10,-12.5 27.3,-13.3l12.3,-0.5 -0.2,-17.6 -0.3,-17.6 -9,-1.2c-12.3,-1.6 -31.7,-1.5 -38.8,0.3z"/>
</vector>
```

This is how it should look like (rendered in Android Studio):

![image](https://user-images.githubusercontent.com/3057584/147827121-8f4dd0bf-d3d0-43ef-852e-582ca8554557.png)

But the library instead renders an empty icon without failing or printing any error; I have no idea why... could you look into this?

**Comment by zathras** [comment]
Thanks.  The issue is that jovial_svg should have been ignoring the width and height attribute of the AVD file.  They don't specify the width and height of the AVD.  I think they probably specify a default rendering width and height for the user agent, which is irrelevant to this library.

The AVD file syntax is well-specified by Google, but the rendering semantics aren't.

Anyway, for now you can just nuke the width and height attributes in your AVD file.  I'll spin a jovial_svg release relatively soon with this fix in it.

**Comment by alesimula** [comment]
@zathras Thanks, I think the width/height aspect ratio is probably all that matters

![image](https://user-images.githubusercontent.com/3057584/147875352-d3d00f3b-b1ba-446e-90d4-983595f52f64.png)
^ Icon on android studio with width = (1/2) * height

![image](https://user-images.githubusercontent.com/3057584/147875379-68387b19-3f96-472b-8be5-02256429fc5f.png)
^ Icon on android studio with height = (1/2) * width

**Comment by zathras** [comment]
Good point.  A 100% complete implementation of android vector drawables would need to do that.

I looked into it, and it would be pretty tricky to pull off.  But more to the point, I don't think this is a very important thing to support for AVDs - AFAIK AVDs are mostly useful for the little icons that come with the Android environment, and other, similar things that would never use scaling in this way.  I could be wrong, and I'd be happy to reconsider if you or someone else could show me someone actually using this parameter in this way, for something non-contrived...  But if someone is authoring assets, the under-specified, hacky AVD format would be an odd choice, given how much more capable, specified and well-supported SVG is.

Changing the aspect ratio like that is a weird, limited subset of what one can do with a proper transform matrix.  In SVG, for example, you could just put a group with a transform at the top of the tree, and do scaling *and* translation (and shear, for that matter).  Just doing scale without the ability to do translation is... of limited value.

About it being tricky:  The thing is, AFAIK AVD files aren't required to specify a viewportWidth and viewportHeight.  But you need to know that, in order to know the scale factor.  That means that you potentially have to parse the whole AVD, calculate the viewport size (the extent of the AVDs drawing), then compare that to the scaled width and height, and apply any needed scaling factor.  Applying a scaling factor to an SI instance would be non-trivial.  I question if it would be worth it, especially since SVGs are so much more capable, and are the main target of this library.

But this limitation should be documented.  I'll put it in the README, with a link to this issue.

**Comment by alesimula** [comment]
@zathras Provided I know nothing about rendering vector icons, let's assume the android:width is half of the height, can't you just divide the viewportHeight by 2 along with the y coordinate of each point?

**Comment by alesimula** [comment]
@zathras Thanks. As you've said I've yet to encounter an apk icon (what I use this library for) that makes use of this aspect-ratio feature, so I guess 100% compliancy would have little benefit.

**Comment by zathras** [comment]
@alesimula About dividing by two, etc...  Yes, running down the tree and transforming individual points would be the starting point for one way of implementing it.  Even that would take some care to get right, since there are two internal representations (DAG tree and compact), and they both canonicalize identical subtrees down to a single subtree (hence "directed acyclic graph," and not "tree").  Then, if AVD has outline width (it does, doesn't it?  I'd have to check to be sure), something would have to be done about that...  I assume that the AVD behavior in that case would be the same as a scaling transform, that is, horizontal lines would come out a different thickness than vertical lines, and angled lines somewhere in between.

Oh, wait, assuming that's true, then the transform-points-down-the-graph approach would be impractical, and the only choice would be to put a scaling transform at the top of the tree.  That's certainly doable.  In the in-memory DAG representation it's at least straightforward; in the compact representation, it turns into a bit of a mess.  It is possible, but it's an awful lot of trouble for a little-used (and maybe never used) feature of a secondary format for the library.

**Comment by zathras** [comment]
I guess I'm too much of an obsessive-compulsive software engineering type to leave that undone.  I really doubt it's of value, but implementing AVD scaling in the (normal? required?) case where viewportWidth and viewportHeight are specified does no harm.

**Comment by alesimula** [comment]
@zathras lol I feel ya

**Comment by zathras** [comment]
Released in 1.0.8
